pkg/assets/builder.go

```diff
@@ -127,32 +142,119 @@ func (a *AssetBuilder) RemapImage(image string) (string, error) {
-func (a AssetBuilder) RemapFile(file string) (string, error) {
-	if file == "" {
-		return "", fmt.Errorf("unable to remap an empty string")
+func (a *AssetBuilder) RemapFileAndSHA(fileURL *url.URL) (*url.URL, *hashing.Hash, error) {
+	if fileURL == nil {
+		return nil, nil, fmt.Errorf("unable to remap an empty URL")
 	}
 
 	fileAsset := &FileAsset{
-		File:              file,
-		CanonicalLocation: file,
+		FileURL: fileURL,
 	}
 
 	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
-		fileURL, err := url.Parse(file)
+		fileAsset.CanonicalFileURL = fileURL
+
+		normalizedFileURL, err := a.normalizeURL(fileURL)
 		if err != nil {
-			return "", fmt.Errorf("unable to parse file url %q: %v", file, err)
+			return nil, nil, fmt.Errorf("unable to parse file url %q: %v", normalizedFileURL, err)
 		}
 
-		fileRepo := strings.TrimSuffix(*a.AssetsLocation.FileRepository, "/")
-		fileAsset.File = fileRepo + fileURL.Path
+		fileAsset.FileURL = normalizedFileURL
+
+		glog.V(4).Infof("adding remapped file: %+v", fileAsset)
 	}
 
+	h, err := a.findHash(fileAsset)
+	if err != nil {
+		return nil, nil, err
+	}
+	fileAsset.SHAValue = h.Hex()
+
 	a.FileAssets = append(a.FileAssets, fileAsset)
+	glog.V(8).Infof("adding file: %+v", fileAsset)
+
+	return fileAsset.FileURL, h, nil
+}
+
+// TODO - remove this method as CNI does now have a SHA file
+
+// RemapFileAndSHAValue is used exclusively to remap the cni tarball, as the tarball does not have a sha file in object storage.
+func (a *AssetBuilder) RemapFileAndSHAValue(fileURL *url.URL, shaValue string) (*url.URL, error) {
+	if fileURL == nil {
+		return nil, fmt.Errorf("unable to remap an empty URL")
+	}
+
+	fileAsset := &FileAsset{
+		FileURL:  fileURL,
+		SHAValue: shaValue,
+	}
+
+	if a.AssetsLocation != nil && a.AssetsLocation.FileRepository != nil {
+		fileAsset.CanonicalFileURL = fileURL
+
+		normalizedFile, err := a.normalizeURL(fileURL)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse file url %q: %v", normalizedFile, err)
+		}
+
+		fileAsset.FileURL = normalizedFile
+		glog.V(4).Infof("adding remapped file: %q", fileAsset.FileURL.String())
+	}
+
+	a.FileAssets = append(a.FileAssets, fileAsset)
+
+	return fileAsset.FileURL, nil
+}
+
+// FindHash returns the hash value of a FileAsset.
+func (a *AssetBuilder) findHash(file *FileAsset) (*hashing.Hash, error) {
+	u := file.FileURL
+
+	// If the phase is "assets" we use the CanonicalFileURL,
+	// but during other phases we use the hash from the FileRepository or the base kops path.
+	// We do not want to just test for CanonicalFileURL as it is defined in
+	// other phases, but is not used to test for the SHA.
+	// This prevents a chicken and egg problem where the file is not yet in the FileRepository.
+	// Summary
+	// assets phase -> get the sha file from the source / CanonicalFileURL
+	// any other phase -> get the sha file from the kops base location or the FileRepository
```

Comment:
Because of which SHA we want to access. We do not have a SHA @ the FileURL location during the assets phase.  So we need to pull down the CanonicalFileURL SHA, and validate that.  But during any other phase we need to validate the FileURL SHA.  This code works, and using the CanonicalFileURL during cluster phase is invalid.  Cluster phase needs to fail if the FileURL sha does not exist.

Reply:
So this isn't about correct or not, it's about explaining why the code is doing what it is doing.

I do remain unconvinced that we shouldn't use the CanonicalFileURL hash always.  My thoughts: We want to know the SHA of the source file.  The source file is at CanonicalFileURL.  We then need to validate that the FileURL has that SHA, and if not we copy from CanonicalFileURL (or error/whatever)

But, if "using the CanonicalFileURL during cluster phase is invalid", then write why in the comment - that's probably the piece I'm missing.  It's more obvious to you today because you wrote the code, but in a year it won't be as obvious - even to you :-)